Solved problems

IliutAndrei opened the pull request:
@alexandru-calinoiu @levelboy

IliutAndrei committed:
Added gitignore

IliutAndrei committed:
Added test for array sum

IliutAndrei committed:
Added solution for array sum

IliutAndrei committed:
Add test for array with no elements

IliutAndrei committed:
Add test for nil array

IliutAndrei committed:
Add solution for nil array

Refactored the funtion, if the given array is nil it will return 0 to maintain a good design

IliutAndrei committed:
Add test first two largest number for array

IliutAndrei committed:
Add solution for 'first 2 largest numbers' test for problem 2

IliutAndrei committed:
Add test 'first two largest number' for empty array

IliutAndrei committed:
Add test 'first 2 largest numbers from nil array' problem 2

IliutAndrei committed:
Add solution for 'first 2 largest numbers from nil array' test problem2

Refactored the funtion, if the given array is nil it will return [] (empty array) to maintain a good design

IliutAndrei committed:
Add test array 'contains duplicates' problem 3

IliutAndrei committed:
Add solution array 'contains duplicates' problem 3

IliutAndrei committed:
Add test array 'does not contain duplicates' problem 3

IliutAndrei committed:
Add test 'empty array does not contain duplicates' problem 3

IliutAndrei committed:
Add test 'nil array does not contain duplicates' problem 3

IliutAndrei committed:
Add solution 'empty array does not contain duplicates' problem 3

Refactored the funtion, if the given array is nil it will return false to maintain a good design

alexandru-calinoiu commented on src/problem1.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem1
+  def sum(array)
+    unless array.nil?
```
Very c# like syntax here can you write this with no branching?

_Hint: `[].sum == 0` and `conditional assignment`_

alexandru-calinoiu commented on src/problem1.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem1
+  def sum(array)
+    unless array.nil?
+      array.sum
```
The description of the problem said: `compute the sum of all __even__ numbers in this array`

IliutAndrei replied:
Yes indeed, sorry, looks like i need new glasses, I will solve the problem asap.

alexandru-calinoiu commented on src/problem2.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem2
+  def identify_largest_numbers(array, no_elements_identify)
```
The solution asked for the first 2 numbers, no generic implementation was required.

alexandru-calinoiu commented on src/problem2.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem2
+  def identify_largest_numbers(array, no_elements_identify)
+    unless array.nil?
```
Same as above see if you can drop the branching, branch prediction is a hard optimization don't make the processors live hard.

IliutAndrei replied:
Then how do i check if the array is nil? if the function receives a nil parameter is will surely fail. Shouldn't I treat that case in the function?

alexandru-calinoiu commented on src/problem2.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem2
+  def identify_largest_numbers(array, no_elements_identify)
+    unless array.nil?
+      array.max(no_elements_identify)
```
Clever, but the problem required you not to sort the array, if you read the [docs](http://ruby-doc.org/core-2.4.1/Enumerable.html#method-i-max) max will sort the array.

Please implement this without sorting.

alexandru-calinoiu commented on src/problem3.rb:
```diff
@@ -0,0 +1,9 @@
+class Problem3
+  def contains_duplicates(array)
+    unless array.nil?
+      array.uniq.length != array.length
```
If you browse the [docs](http://ruby-doc.org/core-2.4.1/Enumerable.html#method-i-max) and you will click on toggle source you will see that `uniq` implementation uses a hash, the requirement of the problem is to use no additional data structures.

alexandru-calinoiu commented on test/problem1_spec.rb:
```diff
@@ -0,0 +1,24 @@
+
+describe 'Problem1' do
+
+  array = [2,1,3,3,4]
```
This is not the best way to write this variables,

Consider using [let](https://relishapp.com/rspec/rspec-core/docs/helper-methods/let-and-let)

alexandru-calinoiu commented on test/problem1_spec.rb:
```diff
@@ -0,0 +1,24 @@
+  empty_array = []
+  nil_array = nil
+
+  it 'calculate sum' do
```
One of the points of `rspec` is to be able to read the specs and make sens of them.

Take a look at this alternative way of writing them, see what happens when a spec fails:

```ruby
describe SumEven do
  subject { SumEven.new.calculate(array) }

  context 'when array is empty' do
    let(:array) { [] }

    # Bonus points if you can explain how this works?
    it { is_expected.to_be zero }
  end

  context 'when array is nil'

  context 'when array is has even numbers' do
    let(:array) { [1, 2, 4, 36, 3 ] }

    it { is_expected.to eq 42 }
  end

  context 'when array has no even numbers'
end
```

_Note: This is an example unrelated to the current spec_

alexandru-calinoiu reviewed (commented):
Good start, you get points for solving the problems not for being fast.

Read the text of the problems carefully and understand that ruby's magic comes with a cost.

IliutAndrei committed:
Add test 'compute array with even numbers' problem 1

IliutAndrei committed:
Add solution for 'compute array with even numbers' test problem 1

IliutAndrei committed:
Add test 'when array doesnt have even numbers' problem 1

IliutAndrei committed:
Add solution for 'when array doesnt have even numbers' test problem 1

IliutAndrei committed:
Add test 'when array is nil' problem 1

IliutAndrei committed:
Add solution for 'when array is nil' test problem 1

IliutAndrei committed:
Add test for 'when array is empty' problem 1

IliutAndrei committed:
Refactored problem 2 to meet the requirments

The problem is solved without using any sorting method now.

+Added new test 'when array has first two largest numbers equal'

IliutAndrei committed:
Add test 'when array has only 1 number' problem 1

IliutAndrei committed:
Refactored problem 3 to meet the requirments

IliutAndrei commented:
Ultimul commit nu e bun (dupa o recitire a cerintei), il voi reface.

IliutAndrei committed:
Refactored problem 3 to meet the requirments

alexandru-calinoiu commented on src/problem1.rb:
```diff
@@ -0,0 +1,7 @@
+class Problem1
+  def even_numbers_sum(array)
```
If you know that `[].sum == 0` how will this change your solution?

IliutAndrei replied:
In case the array is nil, i have to return 0, otherwise the 'select' function does not take nil as a parameter => error. And to return nil from a function is bad.

alexandru-calinoiu commented on src/problem3.rb:
```diff
@@ -0,0 +1,12 @@
+    return false if array.nil? or array.empty?
+
+    array_length_sum = 0;
+    (0..array.length).each do |n|
```
:) the c# in you is strong, ruby has intervals, one can write`(0..array.lenght).sum`

IliutAndrei replied:
I will refactor that.

alexandru-calinoiu commented on test/problem3_spec.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  context 'when array contains only duplicates' do
+    let(:array) { [1,1,1,1] }
```
Try to change this to `let(:array) { [1, 1, 1] }`

IliutAndrei replied:
Done

alexandru-calinoiu commented on test/problem1_spec.rb:
```diff
@@ -0,0 +1,37 @@
+require 'rspec'
+require_relative '../src/problem1'
+
+describe Problem1 do
```
Specs look much better :+1:

alexandru-calinoiu reviewed (changes requested):
Better, problem 3 is not yet solved, try one more time.

IliutAndrei committed:
Refactored problem 3

IliutAndrei committed:
Refactored problem 3 to pass all tests

IliutAndrei commented:
Acuma testele sunt ok si cu noul array de test [1,1,1] la problema 3.

pirvudoru commented on src/problem2.rb:
```diff
@@ -0,0 +1,18 @@
+class Problem2
+  def identify_largest_numbers(array)
+    return [] if array.nil? or array.empty?
```
Is an empty array a valid output for this function? By name it has to return largest (2) numbers. In this case it cannot accomplish it's goal because some preconditions are not met.

IliutAndrei replied:
Indeed, refactoring now.

pirvudoru commented on test/problem1_spec.rb:
```diff
@@ -0,0 +1,37 @@
+  context 'when array is nil' do
+    let(:array) { nil }
+
+    it { is_expected.to eq 0}
```
Does not seem fair to me. The problem states that you have an array of size N, so it won't be null. 
Also sum of nil to equal 0 seems odd.

pirvudoru commented on test/problem3_spec.rb:
```diff
@@ -0,0 +1,36 @@
+  context 'when array is nil' do
+    let(:array) { nil }
+
+    it { is_expected.to be false }
```
Why guard against nils ? Problems state that you are given an array of size N

pirvudoru commented on test/problem2_spec.rb:
```diff
@@ -0,0 +1,30 @@
+  context 'when array is nil' do
+    let(:array) { nil }
+
+    it { is_expected.to eq [] }
```
Why guard against nils ? Problems state that you are given an array of size N

IliutAndrei replied:
Indeed, I will remove the nil check

pirvudoru commented on src/problem1.rb:
```diff
@@ -0,0 +1,7 @@
+  def even_numbers_sum(array)
+    return 0 if array.nil?
+    sum = array.select(&:even?).reduce(:+)
+    sum == nil ? 0 : sum
```
This can be furthure simplified. Check @alexandru-calinoiu's comment. Also no need to guard for nil. Problems state that you are given an array of size N

IliutAndrei replied:
Done

pirvudoru commented on test/problem3_spec.rb:
```diff
@@ -0,0 +1,36 @@
+require 'rspec'
+require_relative '../src/problem3'
+
+describe Problem3 do
```
Add a case for `[1,1,4,4]`

IliutAndrei replied:
Yes, the test fails, I will try to come up with a better implementation for this problem.

IliutAndrei committed:
Refactored all problems

alexandru-calinoiu commented on src/problem2.rb:
```diff
@@ -0,0 +1,18 @@
+  def identify_largest_numbers(array)
+    raise "Array of length too small, min 2!" if array.length < 2
+
+    max1, max2 = -1
```
This is assuming that there are no negative numbers in the array, `integers` can have negative values.

alexandru-calinoiu reviewed (approved):
Problem 2 still had an issue

alexandru-calinoiu commented:
Will close this and followup on email.
